src/hierarchies/fa_hierarchy.cc

```diff
@@ -0,0 +1,308 @@
+#include "ls_state.pb.h"
+#include "src/utils/proto_utils.h"
+#include "src/utils/rng.h"
+
+double FAHierarchy::like_lpdf(const Eigen::RowVectorXd& datum) const {
+  using stan::math::NEG_LOG_SQRT_TWO_PI;
+  double base = (Eigen::MatrixXd(state.cov_chol.matrixL()))
+                    .diagonal()
+                    .array()
+                    .log()
+                    .sum() -
+                NEG_LOG_SQRT_TWO_PI * dim;
+  double exp =
+      0.5 * ((datum.transpose() - state.mu)
+                 .dot(state.cov_chol.solve((datum.transpose() - state.mu))));
+  return -(base + exp);
+}
+
+FA::State FAHierarchy::draw(const FA::Hyperparams& params) {
+  auto& rng = bayesmix::Rng::Instance().get();
+  FA::State out;
+  out.mu = params.mutilde;
+  out.psi = params.beta / (params.alpha0 + 1.);
+  out.eta = Eigen::MatrixXd::Zero(card, params.q);
+  out.lambda = Eigen::MatrixXd::Zero(dim, params.q);
+
+  for (size_t j = 0; j < dim; j++) {
+    out.mu[j] =
+        stan::math::normal_rng(params.mutilde[j], sqrt(params.phi), rng);
+
+    out.psi[j] = stan::math::inv_gamma_rng(params.alpha0, params.beta[j], rng);
+
+    for (size_t i = 0; i < params.q; i++) {
+      out.lambda(j, i) = stan::math::normal_rng(0, 1, rng);
+    }
+  }
+
+  for (size_t i = 0; i < card; i++) {
+    for (size_t j = 0; j < params.q; j++) {
+      out.eta(i, j) = stan::math::normal_rng(0, 1, rng);
+    }
+  }
+
+  out.psi_inverse = out.psi.cwiseInverse().asDiagonal();
+  out.cov_chol = (out.lambda * out.lambda.transpose() +
+                  Eigen::MatrixXd(out.psi.asDiagonal()))
+                     .llt();
+
+  return out;
+}
+
+void FAHierarchy::initialize_state() {
+  std::cout << "init state" << std::endl;
+  state.mu = hypers->mutilde;
+  state.psi = hypers->beta / (hypers->alpha0 + 1.);
+  state.eta = Eigen::MatrixXd::Zero(card, hypers->q);
+  state.lambda = Eigen::MatrixXd::Zero(dim, hypers->q);
+  state.psi_inverse = state.psi.cwiseInverse().asDiagonal();
+  state.cov_chol = (state.lambda * state.lambda.transpose() +
+                    Eigen::MatrixXd(state.psi.asDiagonal()))
+                       .llt();
+}
+
+void FAHierarchy::initialize_hypers() {
+  if (prior->has_fixed_values()) {
+    // Set values
+    hypers->mutilde = bayesmix::to_eigen(prior->fixed_values().mutilde());
+    dim = hypers->mutilde.size();
+    hypers->beta = bayesmix::to_eigen(prior->fixed_values().beta());
+    hypers->phi = prior->fixed_values().phi();
+    hypers->alpha0 = prior->fixed_values().alpha0();
+    hypers->q = prior->fixed_values().q();
+
+    // Automatic initialization
+    if (dim == 0) {
+      std::cout << "No mutilde found. Initializing with mean." << std::endl;
+      hypers->mutilde = dataset_ptr->colwise().mean();
+      dim = hypers->mutilde.size();
+    }
+    if (hypers->beta.size() == 0) {
+      std::cout << "No beta found. Initializing with scaled precision matrix "
+                   "diagonal."
+                << std::endl;
```

Comment:
Again, I'd rather have these remarks in the documentation than being printed out at runtime in order not to clutter the output